### ISSUE TITLE
Compiler error message is confusing

### ISSUE THREAD

**Issue by divinerapier** [rust-lang/rust#64829]
I don't know what caused this error. Sorry, I can't provide simple code to reproduce it.

![rust-compiler-error](https://user-images.githubusercontent.com/21213088/65736836-0b308900-e10f-11e9-9a1a-c6ef879fbc4a.jpg)

``` bash
$ rustup show
Default host: x86_64-apple-darwin

installed toolchains
--------------------

stable-x86_64-apple-darwin
nightly-2019-09-13-x86_64-apple-darwin (default)
```

Thanks.

**Comment by csmoe** [comment]
You will benefit from reading the [book](https://doc.rust-lang.org/book/ch18-01-all-the-places-for-patterns.html?highlight=misma#let-statements) first.

**Comment by divinerapier** [comment]
@csmoe thank you for your reply. I mean the error message is useless. It says the types mismatched, but there is no context in the error message, so how can I fix it. I know how to fix but I don't know where to fix.

**Comment by csmoe** [comment]
@divinerapier sorry, reopen.

**Comment by divinerapier** [comment]
@csmoe thank you very much

**Comment by Centril** [comment]
Unfortunately, this is report is not actionable without knowing more about the situation, which we will need to know about to fashion a more tailored error message.

**Comment by eddyb** [comment]
@Centril That's not the problem, the problem is there's no span at all.

@divinerapier Are you using any custom derives? That's the only thing I can think of that could be able to cause this. If you can provide a list of your (`proc-macro`) dependencies, that could help a lot.

If you want to try to reduce your code, you can start by taking various functions and replacing their bodies with `unimplemented!()`. If at any point the error goes away, that's the culprit. Otherwise, you can now try to remove various methods and types.

**Comment by estebank** [comment]
This is a problem that needs to be solved. I'm pretty sure it's a duplicate of https://github.com/rust-lang/rust/issues/51635.

**Comment by divinerapier** [comment]
@eddyb When I reduce some code, I find when `function_name::named` is deleted, the error message will be correct. I hope this will help.

[crate function_name](https://crates.io/crates/function_name)

**Comment by steveklabnik** [comment]
Triage: this really needs some way to reproduce.

**Comment by estebank** [comment]
@steveklabnik there have been multiple changes that @Aaron1011 has made to the macro machinery to preserve their token streams to avoid the most egregious of these errors, but they will still happen when proc_macros rewrite the output TTS. I don't recall what the tracking issue for that work is.

**Comment by Aaron1011** [comment]
The tracking issue is https://github.com/rust-lang/rust/issues/43081
I'm currently working on the last major piece (as far as stable code is concerned) in https://github.com/rust-lang/rust/pull/76130.

**Comment by bbros-dev** [comment]
@Aaron1011 given https://github.com/rust-lang/rust/issues/43081 is resolved can this be considered resolved?

I've a confusing compiler error use case to report, and am looking for open issues.

**Comment by Aaron1011** [comment]
@bbros-dev: That's correct - we now preserve spans in all stable code, and in almost all unstable code. Feel free to open a new issue for the compiler error that you're getting.
